#' Summarize results from clustering using a Normal mixture model
#' 
#' A function to summarize the clustering results obtained from a Poisson
#' mixture model.
#' 
#' The summary function for an object of class \code{"PoisMixClus"}
#' provides the number of clusters selected for the BIC, ICL, DDSE, and Djump
#' model selection approaches.
#' 
#' @param object An object of class \code{"NormClus"} 
#' @param y_profiles y (\emph{n} x \emph{q}) matrix of observed profiles for \emph{n}
#' observations and \emph{q} variables
#' @param ... Additional arguments
#' @author Andrea Rau
#' @seealso \code{\link{PoisMixClus}}, \code{\link{PoisMixClus_K}}
#' @references Rau, A., Maugis-Rabusseau, C., Martin-Magniette, M.-L., Celeux,
#' G. (2015) Co-expression analysis of high-throughput transcriptome sequencing
#' data with Poisson mixture models. Bioinformatics, doi:
#' 10.1093/bioinformatics/btu845.
#' 
#' Rau, A., Celeux, G., Martin-Magniette, M.-L., Maugis-Rabusseau, C. (2011).
#' Clustering high-throughput sequencing data with Poisson mixture models.
#' Inria Research Report 7786. Available at
#' \url{http://hal.inria.fr/inria-00638082}.
#' @keywords methods
#' @example /inst/examples/PoisMixClus.R
#' @export
summary.NormMixClus <-
  function (object, y_profiles, ...) 
  {
    x <- object
    if (class(x) != "NormMixClus") {
      stop(paste(sQuote("object"), sep = ""), " must be of class ", 
           paste(dQuote("NormMixClus"), sep = ""), sep = "")
    }
    cat("*************************************************\n")
    clustNum <-  paste(x$nbClust.all, collapse=",")
    cat("Clusters fit: ", clustNum , "\n", sep = "")
    cat("Selected number of clusters via ICL: ", x$ICL.results$K, "\n", sep = "")
    cat("*************************************************\n")
    
    x <- object$ICL.results
    
    probaPost <- x$probaPost
    labels <- apply(probaPost, 1, which.max)
    

    
    map <- apply(probaPost, 1, max)
    length(which(map > 0.9))/length(map)
    
    tab <- table(labels)
    names(tab) <- paste("Cluster", names(tab))
    param <- NormMixParam(x, y_profiles) 
    mu <- param$mu
    pi <- param$pi
    rownames(mu) <- names(pi) <- names(tab)
    g <- x$K
    
    cat("Cluster sizes:\n"); print(tab); cat("\n")
    cat("Number of observations with MAP > 0.90 (% of total):\n")
    cat(length(which(map > 0.9)), " (", round(length(which(map > 0.9))/length(map)*100,2),
        "%)\n\n", sep = "")
    cat("Number of observations with MAP > 0.90 per cluster (% of total per cluster):\n"); 
    
    tab2 <- matrix(NA, nrow = 2, ncol = g)
    colnames(tab2) <- paste("Cluster", 1:g); rownames(tab2) <- rep("", 2)
    for(i in 1:g) {
      if(sum(labels == i) > 1) {
        map.clust <- apply(matrix(probaPost[labels == i,], ncol=g), 1, max)
        tab2[1,i] <- length(which(map.clust > 0.9))
        tab2[2,i] <- paste("(", round(100*length(which(map.clust > 0.9))/length(map.clust),2),
                           "%)", sep = "")
      }
      if(sum(labels == i) == 1) {
        map.clust <- max(probaPost[labels == i,])
        tab2[1,i] <- length(which(map.clust > 0.9))
        tab2[2,i] <- paste("(", round(100*length(which(map.clust > 0.9))/length(map.clust),2),
                           "%)", sep = "")
      }
      if(sum(labels == i) == 0) {
        tab2[1,i] <- "---"
        tab2[2,i] <- "---"
      }
    }
    print(tab2, quote = FALSE); cat("\n")
    
    cat("Mu:\n"); print(round(mu,2)); cat("\n")
    cat("Pi:\n"); print(round(pi,2)); cat("\n")
    
    
  }




